clib/feldspar_array.h

```diff
@@ -36,150 +36,86 @@
-    uint32_t bytes;    /* The number of bytes the buffer can hold */
 };
 
 /* Indexing into an array: */
 /* Result: element of type 'type' */
 #define at(type,arr,idx) (((type*)((arr)->buffer))[idx])
 
-/* Array (re)initialization */
-static inline struct array *initArray(struct array *arr, int32_t size, int32_t len)
+/* Allocate and initialize a struct array if we did not have one already. */
+static inline struct array* allocArray( struct array* src )
 {
-    int newBytes;
+  if( src == NULL ) {
+    src = (struct array *) malloc( sizeof(struct array) );
+    src->buffer = NULL;
+    src->length = 0;
+  }
+  return src;
+}
 
-    log_3("initArray %p %d %d - enter\n", arr, size, len);
-    if ( !arr )
-    {
-      struct array * ptr;
-      ptr = calloc(1, sizeof(struct array));
-      log_4("initArray %p %d %d - alloc fresh struct array %p\n", arr, size, len, ptr);
-      arr = ptr;
-    }
+/* Resizing an existing array (the struct array must be initialized). */
+static inline struct array* resizeArray( struct array* arr, int32_t size, int32_t len )
+{
+  assert( arr );
+  arr->buffer = realloc( arr->buffer, len*size );
+  arr->length = len;
+  return arr;
+}
 
-    assert(arr);
-    arr->elemSize = size;
-    arr->length   = len;
-    if( size < 0 )
-        size = sizeof(struct array);
-    newBytes = size * len;
-    if( arr->buffer )
-    {
-        // Re-initialization
-        log_3("initArray %p %d %d - reinitialize\n",arr,size,len);
-        if( arr->bytes < newBytes )
-        {
-            log_5("initArray %p %d %d - realloc since %d < %d\n"
-                 , arr, size, len, arr->bytes, newBytes);
-            // Not enough space: reallocation needed
-            arr->bytes  = newBytes;
-            arr->buffer = realloc(arr->buffer, newBytes);
-        }
-        else
-        {
-            // Otherwise: space is enough, nothing to do
-            log_5("initArray %p %d %d - large enough %d >= %d\n"
-                 , arr, size, len, arr->bytes, newBytes);
-        }
-    }
-    else
-    {
-        // First initialization
-        arr->bytes = newBytes;
-        arr->buffer = (void*)malloc(newBytes);
-        log_5("initArray %p - alloc %d * %d = %d bytes at %p\n"
-             , arr, arr->length, arr->elemSize, newBytes, arr->buffer);
-    }
-    assert( arr->buffer );
-    log_3("initArray %p %d %d - leave\n", arr, size, len);
-    return arr;
+/* Array (re)initialization for flat arrays. */
+static inline struct array* initArray( struct array* src, int32_t size, int32_t len )
+{
+  src = allocArray( src );
+  if( len != src->length ) {
+    src = resizeArray( src, size, len );
+  }
+  return src;
 }
 
-// Free array
-// TODO: Think about arrays escaping from their scope.
-static inline void freeArray(struct array *arr)
+/* Free a flat array or an array where all the arrays it contains have been free'd already. */
+/* TODO: Think about arrays escaping from their scope. */
```

Comment:
We don't support anything older than C99, and that has line comments.

Reply:
and?